sql/core/src/main/scala/org/apache/spark/sql/execution/merge/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,627 @@
+    }
+  }
+
+  // First traversal builds up the cache and inserts `ScalarSubqueryReference`s to the plan.
+  private def insertReferences(plan: LogicalPlan, cache: ArrayBuffer[Header]): LogicalPlan = {
+    plan.transformUpWithSubqueries {
+      case n => n.transformExpressionsUpWithPruning(_.containsAnyPattern(SCALAR_SUBQUERY)) {
+        case s: ScalarSubquery if !s.isCorrelated && s.deterministic =>
+          val (subqueryIndex, headerIndex) = cacheSubquery(s.plan, cache)
+          ScalarSubqueryReference(subqueryIndex, headerIndex, s.dataType, s.exprId)
+      }
+    }
+  }
+
+  // State of the plan merging algorithm
+  object ScanCheck extends Enumeration {
+    type ScanCheck = Value
+
+    // There is no need to check if physical plan is mergeable until we don't encounter `Filter`s
+    // with different predicates
+    val NO_NEED,
+
+    // We switch to this state once we encounter different `Filters` in the plans we want to merge.
+    // Identical logical plans is not enough to consider the plans mergeagle, but we also need to
+    // check physical scans to determine if partition, bucket and other pushed down filters are
+    // safely mergeable without performance degradation.
+    CHECKING,
+
+    DONE = Value
+    // Once the physical check is complete we use this state to finish the logical merge check.
+  }
+
+  import ScanCheck._
+
+  // Caching returns the index of the subquery in the cache and the index of scalar member in the
+  // "Header".
+  private def cacheSubquery(plan: LogicalPlan, cache: ArrayBuffer[Header]): (Int, Int) = {
+    val output = plan.output.head
+    cache.zipWithIndex.collectFirst(Function.unlift { case (header, subqueryIndex) =>
+      checkIdenticalPlans(plan, header.plan).map { outputMap =>
+        val mappedOutput = mapAttributes(output, outputMap)
+        val headerIndex = header.attributes.indexWhere(_.exprId == mappedOutput.exprId)
+        subqueryIndex -> headerIndex
+      }.orElse(tryMergePlans(plan, header.plan, NO_NEED).collect {
+        case (mergedPlan, outputMap, None, None) =>
+          val mappedOutput = mapAttributes(output, outputMap)
+          var headerIndex = header.attributes.indexWhere(_.exprId == mappedOutput.exprId)
+          val newHeaderAttributes = if (headerIndex == -1) {
+            headerIndex = header.attributes.size
+            header.attributes :+ mappedOutput
+          } else {
+            header.attributes
+          }
+          cache(subqueryIndex) = Header(newHeaderAttributes, mergedPlan, true)
+          subqueryIndex -> headerIndex
+      })
+    }).getOrElse {
+      cache += Header(Seq(output), plan, false)
+      cache.length - 1 -> 0
+    }
+  }
+
+  // If 2 plans are identical return the attribute mapping from the new to the cached version.
+  private def checkIdenticalPlans(
+      newPlan: LogicalPlan,
+      cachedPlan: LogicalPlan): Option[AttributeMap[Attribute]] = {
+    if (newPlan.canonicalized == cachedPlan.canonicalized) {
```

Comment:
[doubt] Does this works with V2 sources as well, considering [earlyScanPushDownRules](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala#L37) makes changes to the scan, hence changing the canonicalization of the scalar subqueries ?

Reply:
No this doesn't work with DSv2 sources (nor did the the original https://github.com/apache/spark/pull/32298).

I'm planning to add DSv2 support in another follow-up PR. Probably with introducing an `SupportsMerge` interface that `Scan`s can implement to merge with another `Scan`.

Reply:
I opened https://github.com/apache/spark/pull/37711 to add support for DSv2 sources (only Parquet first).